### ISSUE TITLE
for loop over results gives invalid improvement suggestion

### ISSUE THREAD

**Issue by pronebird** [rust-lang/rust#140401]
I tried this code:

```rs
let mut dir_stream = ReadDirStream::new(tokio::fs::read_dir(data_dir).await.unwrap());

for entry_result in dir_stream.next().await {
    if let Ok(entry) = entry_result {
        tracing::info!("File left in data directory: {}", entry.path().display());
    }
}
```

I expected to see this happen: no warnings

Instead, this happened: warning thrown
```
warning: for loop over an `Option`. This is more readably written as an `if let` statement

138 |     for entry_result in dir_stream.next().await {
    |                         ^^^^^^^^^^^^^^^^^^^^^^^
    |
help: to check pattern in a loop use `while let`
    |
138 -     for entry_result in dir_stream.next().await {
138 +     while let Some(entry_result) = dir_stream.next().await {
    |
help: consider using `if let` to clear intent
    |
138 -     for entry_result in dir_stream.next().await {
138 +     if let Some(entry_result) = dir_stream.next().await {

```
It's strange that Rust suggests to use while loop as it has a different meaning.  The iterator yields `Result<DirEntry>` and not `Option<DirEntry>`.

These are the inline hints in VSCode:

![Image](https://github.com/user-attachments/assets/ef8ea83a-8937-46a8-abbc-881ec535be1e)

### Meta

`rustc --version --verbose`:
```
rustc 1.88.0-nightly (cb31a009e 2025-04-27)
binary: rustc
commit-hash: cb31a009e3e735ab08613cec2d8a5a754e65596f
commit-date: 2025-04-27
host: aarch64-apple-darwin
release: 1.88.0-nightly
LLVM version: 20.1.2
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by pronebird** [comment]
OK nevermind, long day. next().await returns `Option<Result<DirEntry>>` so Rust suggests to use while loop instead. Not sure why this is better though.
